### ISSUE TITLE
Column names incorrect when using a non default delimiter

### ISSUE THREAD

**Issue by dveeden** [pingcap/tidb#39662]
## Bug Report

This is with MySQL Shell 8.0 as client. I would expect that this also happens with MySQL Client.

```
sql> delimiter $$

sql> SELECT 1; SELECT 2; $$
+--------------+
| 1; SELECT 2; |
+--------------+
|            1 |
+--------------+
1 row in set (0.0004 sec)

+----+
| 2; |
+----+
|  2 |
+----+
1 row in set (0.0004 sec)

sql> SELECT 1 AS 'a'; SELECT 2 AS 'b'; $$
+---+
| a |
+---+
| 1 |
+---+
1 row in set (0.0004 sec)

+---+
| b |
+---+
| 2 |
+---+
1 row in set (0.0004 sec)

sql> delimiter ;

sql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.4.0
Edition: Community
Git Commit Hash: cf36a9ce2fe1039db3cf3444d51930b887df18a1
Git Branch: heads/refs/tags/v6.4.0
UTC Build Time: 2022-11-13 05:25:30
GoVersion: go1.19.2
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
1 row in set (0.0003 sec)
```

I expected this to have the same column names as when the default delimiter is used:
```
sql> SELECT 1; SELECT 2;
+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.0006 sec)
+---+
| 2 |
+---+
| 2 |
+---+
1 row in set (0.0002 sec)
```

**Comment by dveeden** [comment]
MySQL 8.0.31 with MySQL Client
```
mysql-8.0.31> delimiter $$
mysql-8.0.31> SELECT 1; SELECT 2; $$
+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.00 sec)

+---+
| 2 |
+---+
| 2 |
+---+
1 row in set (0.00 sec)
```

**Comment by dveeden** [comment]
TiDB v6.4.0 with MySQL Client
```
mysql> delimiter $$
mysql> SELECT 1; SELECT 2; $$
+-------------+
| 1; SELECT 2 |
+-------------+
|           1 |
+-------------+
1 row in set (0.00 sec)

+---+
| 2 |
+---+
| 2 |
+---+
1 row in set (0.00 sec)
```

**Comment by zanmato1984** [comment]
I believe this is parser related issue. Adjusting labels.

**Comment by dveeden** [comment]
This is how this looks in Wireshark:

The query:
![image](https://user-images.githubusercontent.com/1272980/213408708-c0355d26-459b-45be-930a-38dc8c343b7d.png)

First resultset:
![image](https://user-images.githubusercontent.com/1272980/213408872-7dbdacfc-c3a1-45d2-b602-df0104eb5faa.png)

Second resultset:
![image](https://user-images.githubusercontent.com/1272980/213408907-6a1cf324-2a02-4459-b450-389e3b9c2dc8.png)

So basically the `mysql.field.name` is `1; SELECT 2` where we expect `1` instead, just like the client shows.

**Comment by dveeden** [comment]
This indeed looks like it is related to the parser:
```
(dlv) list
> github.com/pingcap/tidb/parser.(*Parser).ParseSQL() ./parser/yy_parser.go:163 (PC: 0x2caad34)
   158:		parser.result = parser.result[:0]
   159:	
   160:		var l yyLexer = &parser.lexer
   161:		yyParse(l, parser)
   162:	
=> 163:		warns, errs := l.Errors()
   164:		if len(warns) > 0 {
   165:			warns = append([]error(nil), warns...)
   166:		} else {
   167:			warns = nil
   168:		}
(dlv) p sql
"SELECT 1; SELECT 2;"
(dlv) p parser.result[0].Fields.Fields[0].node.text
"1; SELECT 2"
(dlv) p parser.result[1].Fields.Fields[0].node.text
"2"
```

**Comment by dveeden** [comment]
Patch to add a test for this:
```diff
diff --git a/parser/parser_test.go b/parser/parser_test.go
index 2fb9b073f..7aee985b0 100644
--- a/parser/parser_test.go
+++ b/parser/parser_test.go
@@ -7142,3 +7142,14 @@ func TestTTLTableOption(t *testing.T) {
 
        RunTest(t, table, false)
 }
+
+func TestMultiStmt(t *testing.T) {
+       p := parser.New()
+       stmts, _, err := p.Parse("SELECT 'foo'; SELECT 'bar'", "", "")
+       require.NoError(t, err)
+       require.Equal(t, len(stmts), 2)
+       stmt1 := stmts[0].(*ast.SelectStmt)
+       stmt2 := stmts[1].(*ast.SelectStmt)
+       require.Equal(t, "'foo'", stmt1.Fields.Fields[0].Text())
+       require.Equal(t, "'bar'", stmt2.Fields.Fields[0].Text())
+}
```

**Comment by dveeden** [comment]
Running the test with parser debugging enabled:
```
$ go test -v -run TestMultiStmt
=== RUN   TestMultiStmt

lex selectKwd(0xe0b4 57524), lval: &{yys:0 offset:0 item:<nil> ident:SELECT expr:<nil> statement:<nil>}
shift, and goto state 44

lex stringLit(0xe007 57351), lval: &{yys:0 offset:7 item:<nil> ident:foo expr:<nil> statement:<nil>}
reduce using rule 1719 (SelectStmtOpts), and goto state 3485
shift, and goto state 730

lex ';'(0x3b 59), lval: &{yys:0 offset:12 item:<nil> ident:; expr:<nil> statement:<nil>}
reduce using rule 1263 (StringLiteral), and goto state 726
reduce using rule 1257 (Literal), and goto state 741
reduce using rule 1305 (SimpleExpr), and goto state 734
reduce using rule 1295 (BitExpr), and goto state 231
reduce using rule 673 (PredicateExpr), and goto state 230
reduce using rule 645 (BoolPri), and goto state 229
reduce using rule 620 (Expression), and goto state 3500
reduce using rule 682 (FieldAsNameOpt), and goto state 3514
reduce using rule 681 (Field), and goto state 3503
reduce using rule 688 (FieldList), and goto state 3504
reduce using rule 1728 (SelectStmtFieldList), and goto state 3505
reduce using rule 691 (HavingClause), and goto state 3506
reduce using rule 1558 (SelectStmtBasic), and goto state 45
reduce using rule 2341 (WhereClauseOptional), and goto state 3461
reduce using rule 1729 (SelectStmtGroup), and goto state 3462
reduce using rule 1278 (OrderByOptional), and goto state 3463
reduce using rule 1708 (SelectStmtLimitOpt), and goto state 3464
reduce using rule 1737 (SelectLockOpt), and goto state 3465
reduce using rule 1731 (SelectStmtIntoOption), and goto state 3466
reduce using rule 1573 (SelectStmt), and goto state 57
reduce using rule 2094 (Statement), and goto state 156
reduce using rule 2140 (StatementList), and goto state 2
shift, and goto state 4482

lex selectKwd(0xe0b4 57524), lval: &{yys:0 offset:14 item:<nil> ident:SELECT expr:<nil> statement:<nil>}
shift, and goto state 44

lex stringLit(0xe007 57351), lval: &{yys:0 offset:21 item:<nil> ident:bar expr:<nil> statement:<nil>}
reduce using rule 1719 (SelectStmtOpts), and goto state 3485
shift, and goto state 730

lex $end(0xe000 57344), lval: &{yys:0 offset:26 item:<nil> ident: expr:<nil> statement:<nil>}
reduce using rule 1263 (StringLiteral), and goto state 726
reduce using rule 1257 (Literal), and goto state 741
reduce using rule 1305 (SimpleExpr), and goto state 734
reduce using rule 1295 (BitExpr), and goto state 231
reduce using rule 673 (PredicateExpr), and goto state 230
reduce using rule 645 (BoolPri), and goto state 229
reduce using rule 620 (Expression), and goto state 3500
reduce using rule 682 (FieldAsNameOpt), and goto state 3514
reduce using rule 681 (Field), and goto state 3503
reduce using rule 688 (FieldList), and goto state 3504
reduce using rule 1728 (SelectStmtFieldList), and goto state 3505
reduce using rule 691 (HavingClause), and goto state 3506
reduce using rule 1558 (SelectStmtBasic), and goto state 45
reduce using rule 2341 (WhereClauseOptional), and goto state 3461
reduce using rule 1729 (SelectStmtGroup), and goto state 3462
reduce using rule 1278 (OrderByOptional), and goto state 3463
reduce using rule 1708 (SelectStmtLimitOpt), and goto state 3464
reduce using rule 1737 (SelectLockOpt), and goto state 3465
reduce using rule 1731 (SelectStmtIntoOption), and goto state 3466
reduce using rule 1573 (SelectStmt), and goto state 57
reduce using rule 2094 (Statement), and goto state 4483
reduce using rule 2141 (StatementList), and goto state 2
reduce using rule 1 (Start), and goto state 1
accept
    parser_test.go:7150: 
        	Error Trace:	parser_test.go:7150
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 4
        	Test:       	TestMultiStmt
--- FAIL: TestMultiStmt (0.00s)
FAIL
exit status 1
FAIL	github.com/pingcap/tidb/parser	0.007s
```

**Comment by dveeden** [comment]
It looks like the `lastEnd` is set to 26 instead of 12.
```
$ dlv test -- -test.run TestMultiStmt
Type 'help' for list of commands.
(dlv) b /home/dvaneeden/dev/pingcap/tidb/parser/parser.go:17568
Breakpoint 1 set at 0x856078 for github.com/pingcap/tidb/parser.yyParse() ./parser.go:17568
(dlv) c
> github.com/pingcap/tidb/parser.yyParse() ./parser.go:17568 (hits goroutine(6):1 total:1) (PC: 0x856078)
  17563:						lastEnd = len(src)
  17564:						if src[lastEnd-1] == ';' {
  17565:							lastEnd--
  17566:						}
  17567:					}
=>17568:					lastField.SetText(parser.lexer.client, src[lastField.Offset:lastEnd])
  17569:				}
  17570:				if yyS[yypt-5].item != nil {
  17571:					st.Where = yyS[yypt-5].item.(ast.ExprNode)
  17572:				}
  17573:				if yyS[yypt-4].item != nil {
(dlv) p src
"SELECT 'foo'; SELECT 'bar'"
(dlv) p lastField.Offset
7
(dlv) p lastEnd
26
(dlv) set lastEnd = 12
(dlv) c
> github.com/pingcap/tidb/parser.yyParse() ./parser.go:17568 (hits goroutine(6):2 total:2) (PC: 0x856078)
  17563:						lastEnd = len(src)
  17564:						if src[lastEnd-1] == ';' {
  17565:							lastEnd--
  17566:						}
  17567:					}
=>17568:					lastField.SetText(parser.lexer.client, src[lastField.Offset:lastEnd])
  17569:				}
  17570:				if yyS[yypt-5].item != nil {
  17571:					st.Where = yyS[yypt-5].item.(ast.ExprNode)
  17572:				}
  17573:				if yyS[yypt-4].item != nil {
(dlv) c
PASS
Process 884391 has exited with status 0
```
